src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -40,73 +153,201 @@ instance TypeHasDoc MigrationStatus where
+
+type StorageC store pm =
+  ( StorageContains store
+    [ "sLedger" := LedgerKey ~> LedgerValue
+    , "sOperators" := (Address, Address) ~> ()
+    , "sTokenAddress" := Address
+    , "sAdmin" := Address
     , "sPendingOwner" := Maybe Address
     , "sLedger" := Ledger
     , "sMigrationStatus" := MigrationStatus
+
+    , "sVotingPeriod" := VotingPeriod
+    , "sQuorumThreshold" := QuorumThreshold
+
+    , "sProposals" := ProposalKey ~> Proposal pm
+    , "sProposalKeyListSortByDate" := [ProposalKey]
     ]
+  )
 
 -- | Parameter of the BaseDAO contract
-data Parameter
+data Parameter proposalMetadata
   = Call_FA2 FA2.Parameter
   | Transfer_ownership TransferOwnershipParam
   | Accept_ownership ()
   | Migrate MigrateParam
   | Confirm_migration ()
+  | Propose (ProposeParams proposalMetadata)
+  | Proposal_metadata (View ProposalKey (Proposal proposalMetadata))
+  | Vote [VoteParam]
+  -- Admin
+  | Set_voting_period VotingPeriod
+  | Set_quorum_threshold QuorumThreshold
+  | Flush ()
   deriving stock (Generic, Show)
-  deriving anyclass (IsoValue)
 
-instance ParameterHasEntrypoints Parameter where
-  type ParameterEntrypointsDerivation Parameter = EpdDelegate
+instance (HasAnnotation pm, NiceParameter pm) => ParameterHasEntrypoints (Parameter pm) where
+  type ParameterEntrypointsDerivation (Parameter pm) = EpdDelegate
+
+deriving anyclass instance (WellTypedIsoValue pm) => IsoValue (Parameter pm)
 
 type TransferOwnershipParam = ("newOwner" :! Address)
 type MigrateParam = ("newAddress" :! Address)
 
-emptyStorage :: Storage
+-- | Voting period in seconds
+type VotingPeriod = Natural
+
+-- | QuorumThreshold that a proposal need to meet
+-- quorum_threshold = upvote - downvote
+type QuorumThreshold = Natural
+
+emptyStorage :: Storage pm
 emptyStorage = Storage
   { sLedger =  BigMap $ Map.empty
   , sOperators = BigMap $ Map.empty
-  -- TODO: make it configurable
   , sTokenAddress = genesisAddress
   , sPendingOwner = Nothing
   , sAdmin = genesisAddress
   , sMigrationStatus = NotInMigration
+  , sVotingPeriod = 60 * 60 * 24 * 7 -- 7 days
+  , sQuorumThreshold = 10 -- upvotes
+
+  , sProposals = BigMap $ Map.empty
+  , sProposalKeyListSortByDate = []
   }
 
 mkStorage
   :: Address
   -> Map (Address, FA2.TokenId) Natural
   -> Operators
-  -> Storage
+  -> Storage pm
 mkStorage admin balances operators = Storage
   { sLedger = BigMap balances
   , sOperators = operators
   , sTokenAddress = genesisAddress
   , sPendingOwner = Nothing
   , sAdmin = admin
   , sMigrationStatus = NotInMigration
+  , sVotingPeriod = 60 * 60 * 24 * 7 -- days
+  , sQuorumThreshold = 10 -- upvotes
+
+  , sProposals = BigMap $ Map.empty
+  , sProposalKeyListSortByDate = []
+  }
+
+------------------------------------------------------------------------
+-- Proposal
+------------------------------------------------------------------------
+
+type ProposalKey = ByteString
+
+data Proposal proposalMetadata = Proposal
```

Comment:
It'd be nice to comment more types added in this module.

Reply:
I added a comment to describe this type as well as update the spec.